test/replication-luatest/recovery_triggers_test.lua

```diff
@@ -0,0 +1,107 @@
+local t = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+
+local g = t.group('gh-5272')
+
+local run_before_cfg = [[
+    rawset(_G, 'test_states', {})
+    box.ctl.on_recovery_state(function(state)
+        table.insert(test_states, state)
+    end)
+]]
+
+g.before_all(function(cg)
+    cg.cluster = cluster:new({})
+    cg.server = cg.cluster:build_and_add_server({
+        alias = 'server',
+        args = {run_before_cfg},
+    })
+    cg.replica = cg.cluster:build_and_add_server({
+        alias = 'replica',
+        box_cfg = {
+            replication = server.build_instance_uri('server'),
+        },
+        args = {run_before_cfg},
```

Comment:
Passing the code like this is unidiomatic. Please add a new properly named argument to the server constructor instead. Also, better to pass the code in an environment variable instead of the argument list, because environment variables have names.

Reply:
Ok, what do you think of this? I don't see a point in a separate named argument, primarily because this feature depends on a specific instance file. So a named environment variable should do:
```diff
diff --git a/test/instances/default.lua b/test/instances/default.lua
index 8deda9d0b..e933f22d8 100755
--- a/test/instances/default.lua
+++ b/test/instances/default.lua
@@ -38,7 +38,7 @@ end
 -- such bad tests hang and fail.
 box.ctl.set_on_shutdown_timeout(TIMEOUT_INFINITY)
 
-local exec_arg = arg[1]
+local exec_arg = os.getenv('TARANTOOL_RUN_BEFORE_CFG')
 if exec_arg then
     local run_before_cfg = loadstring(exec_arg)
     run_before_cfg()
```

```diff
diff --git a/test/replication-luatest/recovery_triggers_test.lua b/test/replication-luatest/recovery_triggers_test.lua
index 870990032..755bf2523 100644
--- a/test/replication-luatest/recovery_triggers_test.lua
+++ b/test/replication-luatest/recovery_triggers_test.lua
@@ -15,14 +15,18 @@ g.before_all(function(cg)
     cg.cluster = cluster:new({})
     cg.server = cg.cluster:build_and_add_server({
         alias = 'server',
-        args = {run_before_cfg},
+        env = {
+            ['TARANTOOL_RUN_BEFORE_CFG'] = run_before_cfg,
+        },
     })
     cg.replica = cg.cluster:build_and_add_server({
         alias = 'replica',
         box_cfg = {
             replication = server.build_instance_uri('server'),
         },
-        args = {run_before_cfg},
+        env = {
+            ['TARANTOOL_RUN_BEFORE_CFG'] = run_before_cfg,
+        },
     })
 end)

```

Reply:
Looks good.